File: Conversoes de Bases/Decimal e Romano/DecimalRomano.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define ALGARISMOS_ROMANOS_COUNT 13
#define LIMITE_MAXIMO_DECIMAL 3999
#define LIMITE_MAXIMO_ROMANO "MMMCMXCIX"

const char* algarismosRomanos[ALGARISMOS_ROMANOS_COUNT] = {"I", "IV", "V", "IX", "X", "XL", "L", "XC", "C", "CD", "D", "CM", "M"};
const int valoresDecimais[ALGARISMOS_ROMANOS_COUNT] = {1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000};

char* DecimalParaRomano(int decimal) {
    char romano[20] = "";
    int i = ALGARISMOS_ROMANOS_COUNT - 1;
    while (decimal > 0) {
        if (decimal >= valoresDecimais[i]) {
            strcat(romano, algarismosRomanos[i]);
            decimal -= valoresDecimais[i];
        }
        else {
            i--;
        }
    }
    return romano;
}

int RomanoParaDecimal(const char* romano) {
    int resultado = 0;
    int valorAnterior = 0;
    int tamanhoRomano = strlen(romano);
    for (int i = tamanhoRomano - 1; i >= 0; i--) {
        int valorAtual = 0;
        int valorEncontrado = 0;
        for (int j = 0; j < ALGARISMOS_ROMANOS_COUNT; j++) {
            if (toupper(romano[i]) == *algarismosRomanos[j]) {
                valorAtual = valoresDecimais[j];
                valorEncontrado = 1;
                break;
            }
        }
        if (!valorEncontrado) {
            resultado = 0;
            break;
        }
        if (valorAtual < valorAnterior) {
            resultado -= valorAtual;
        }
        else {
            resultado += valorAtual;
            valorAnterior = valorAtual;
        }
    }
    return resultado;
}

int main(void) {
    int opcao;
    int decimal;
    char romano[20];
    do {
        printf("==========================\n");
        printf(" CONVERSOR DECIMAL-ROMANO \n");
        printf("==========================\n\n");
        printf("1 - Converter decimal para romano\n");
        printf("2 - Converter romano para decimal\n");
        printf("0 - Sair\n\n");
        printf("Digite a opcao desejada: ");
        scanf("%d", &opcao);
        switch (opcao) {
            case 1:
                printf("\nDigite um numero decimal (ate %d): ", LIMITE_MAXIMO_DECIMAL);
                scanf("%d", &decimal);
                if (decimal < 1 || decimal > LIMITE_MAXIMO_DECIMAL) {
                    printf("Valor decimal fora do limite permitido.\n");
                }
                else {
                    printf("Valor em algarismo romano: %s\n", DecimalParaRomano(decimal));
                }
                break;
            case 2:
                printf("\nDigite um numero romano (ate %s): ", LIMITE_MAXIMO_ROMANO);
                scanf("%s", romano);
                if (strlen(romano) == 0 || RomanoParaDecimal(romano) == 0) {
                    printf("Valor romano fora do limite permitido.\n");
                }
                else {
                    printf("Valor em decimal: %d\n", RomanoParaDecimal(romano));
                }
                break;
        }
        printf("\n");
    } while (opcao != 0);
    return 0;
}

File: Conversoes de Bases/Decimal e Romano/DecimalRomano.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

var (
	algarismosRomanos = []string{"I", "IV", "V", "IX", "X", "XL", "L", "XC", "C", "CD", "D", "CM", "M"}
	valoresDecimais   = []int{1, 4, 5, 9, 10, 40, 50, 90, 100, 400, 500, 900, 1000}
	limiteMaximoDecimal = 3999
	limiteMaximoRomano = "MMMCMXCIX"
)

func decimalParaRomano(decimal int) string {
	romano := ""
	i := len(algarismosRomanos) - 1
	for decimal > 0 {
		if decimal >= valoresDecimais[i] {
			romano += algarismosRomanos[i]
			decimal -= valoresDecimais[i]
		} else {
			i--
		}
	}
	return romano
}

func romanoParaDecimal(romano string) int {
	romano = strings.ToUpper(romano)
	resultado := 0
	valorAnterior := 0
	valorEncontrado := false
	for i := len(romano) - 1; i >= 0; i-- {
		valorEncontrado = false
		for valorAtual := 0; valorAtual < len(algarismosRomanos); valorAtual++ {
			if string(romano[i]) == algarismosRomanos[valorAtual] {
				valorEncontrado = true
				if valorAtual < valorAnterior {
					resultado -= valoresDecimais[valorAtual]
				} else {
					resultado += valoresDecimais[valorAtual]
					valorAnterior = valorAtual
				}
				break
			}
		}
		if !valorEncontrado {
			resultado = 0
			break
		}
	}
	return resultado
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Println("==========================")
		fmt.Println(" CONVERSOR DECIMAL-ROMANO ")
		fmt.Println("==========================")
		fmt.Println()
		fmt.Println("1 - Converter decimal para romano")
		fmt.Println("2 - Converter romano para decimal")
		fmt.Println("0 - Sair")
		fmt.Println()
		fmt.Print("Digite a opcao desejada: ")
		scanner.Scan()
		opcao := scanner.Text()
		switch opcao {
		case "1":
			fmt.Println()
			fmt.Printf("Digite um numero decimal (ate %d): ", limiteMaximoDecimal)
			scanner.Scan()
			input := scanner.Text()
			var decimal int
			_, err := fmt.Sscanf(input, "%d", &decimal)
			if err != nil || decimal < 1 || decimal > limiteMaximoDecimal {
				fmt.Println("Valor decimal fora do limite permitido.")
			} else {
				fmt.Printf("Valor em algarismo romano: %s\n", decimalParaRomano(decimal))
			}
		case "2":
			fmt.Println()
			fmt.Printf("Digite um numero romano (ate %s): ", limiteMaximoRomano)
			scanner.Scan()
			romano := scanner.Text()
			if len(romano) == 0 || romanoParaDecimal(romano) == 0 {
				fmt.Println("Valor romano fora do limite permitido.")
			} else {
				fmt.Printf("Valor em decimal: %d\n", romanoParaDecimal(romano))
			}
		case "0":
			return
		}
		fmt.Println()
	}
}
